### PR TITLE
*: add support for IP based bootstrap endpoint

### PR DESCRIPTION
**Pull request by hexfusion**
This PR adds support for etcd-bootstrap to have a client address that is IP vs FQDN. The reason is that customers do not want to add additional DNS records as a dependency for building clusters.

### CODE REVIEW

File: pkg/operator/hostetcdendpointcontroller/members.go [hunk 1]
```diff
@@ -41,19 +43,31 @@ func (h *healthyEtcdMemberGetter) GetHealthyEtcdMembers() ([]string, error) {
...
+			return "", err
+		}
+		host, port, _ := net.SplitHostPort(u.Host)
+		//TODO peer port should be a global constant
+		if isIP(host) && port == "2380" {
+			return "etcd-bootstrap", nil
 		}
 	}
-	return ""
+	return "", nil
```
Comment (hexfusion) [inline pkg/operator/hostetcdendpointcontroller/members.go]: is this an error? not found?

**Comment by hexfusion** [comment]
/retest 

etcd came up ..

File: pkg/operator/hostetcdendpointcontroller/members.go (hunk 1 above)
Comment (alaypatel07) [reply pkg/operator/hostetcdendpointcontroller/members.go]: yes, reaching that point is very unhealthy. I would think a lot of things are not going right for us to reach that state, it would be fair to panic at that point?

File: pkg/operator/hostetcdendpointcontroller/members.go [hunk 2]
```diff
@@ -41,19 +43,31 @@ func (h *healthyEtcdMemberGetter) GetHealthyEtcdMembers() ([]string, error) {
...
 	}
 	return hostnames, nil
 }
 
-func getEtcdName(peerURLs []string) string {
+func getEtcdName(peerURLs []string) (string, error) {
 	for _, peerURL := range peerURLs {
 		if strings.Contains(peerURL, "etcd-") {
-			return strings.TrimPrefix(strings.Split(peerURLs[0], ".")[0], "https://")
+			return strings.TrimPrefix(strings.Split(peerURLs[0], ".")[0], "https://"), nil
```
Comment (alaypatel07) [inline pkg/operator/hostetcdendpointcontroller/members.go]: this existed in old code, my bad in not catching it earlier, but can we do?
```
strings.TrimPrefix(strings.Split(peerURL, ".")[0], "https://"), nil
```

File: pkg/operator/hostetcdendpointcontroller/members.go [hunk 3]
```diff
@@ -163,3 +177,11 @@ func (h *healthyEtcdMemberGetter) Endpoints() ([]string, error) {
 
 	return endpoints, nil
 }
+
```
Comment (alaypatel07) [inline pkg/operator/hostetcdendpointcontroller/members.go]: question: would it be fair to add a TODO here to util this function for using it across the projects?

**Comment by alaypatel07** [comment]
Few of small things:
1. we probably need to change IP address in discovery container, https://github.com/openshift/machine-config-operator/commit/a56735d1ede127255a774a0caa79310146be0c15#diff-f26a1d5dc0cdd6c1f8863a9a3d7a3630R177
2. we should also change the storage URLs as per https://github.com/openshift/cluster-kube-apiserver-operator/pull/673/files, although it will not break things, for consistency
3. squash the commits into a single logical commit?

File: pkg/operator/hostetcdendpointcontroller/members.go (hunk 3 above)
Comment (hexfusion) [reply pkg/operator/hostetcdendpointcontroller/members.go]: agreed

**Comment by hexfusion** [comment]
> Few of small things:
> 
> 1. we probably need to change IP address in discovery container, [openshift/machine-config-operator@a56735d#diff-f26a1d5dc0cdd6c1f8863a9a3d7a3630R177](https://github.com/openshift/machine-config-operator/commit/a56735d1ede127255a774a0caa79310146be0c15#diff-f26a1d5dc0cdd6c1f8863a9a3d7a3630R177)
> 2. we should also change the storage URLs as per https://github.com/openshift/cluster-kube-apiserver-operator/pull/673/files, although it will not break things, for consistency
> 3. squash the commits into a single logical commit?

these will be addressed

File: pkg/operator/hostetcdendpointcontroller/members.go (hunk 2 above)
Comment (alaypatel07) [reply pkg/operator/hostetcdendpointcontroller/members.go]: s/peerURLs[0]/peerURL/

Comment (hexfusion) [reply pkg/operator/hostetcdendpointcontroller/members.go]: yeah gotcha

File: pkg/operator/configobservation/etcd/observe_etcd.go [hunk 4]
```diff
@@ -369,9 +369,9 @@ func (e *etcdObserver) setBootstrapMember() error {
 	for _, subset := range endpoints.Subsets {
 		for _, address := range subset.Addresses {
 			if address.Hostname == "etcd-bootstrap" {
-				name := address.Hostname
-				peerURLs := fmt.Sprintf("https://%s.%s:2380", name, e.ClusterDomain)
-				clusterMember, err := setMember(name, []string{peerURLs}, ceoapi.MemberUnknown)
+				ip := address.IP
```
Comment (alaypatel07) [inline pkg/operator/configobservation/etcd/observe_etcd.go]: drop the extra variable?

Comment (hexfusion) [reply pkg/operator/configobservation/etcd/observe_etcd.go]: it's used more than once I think its fine IMO

File: pkg/operator/hostetcdendpointcontroller/members.go (hunk 3 above)
Comment (hexfusion) [reply pkg/operator/hostetcdendpointcontroller/members.go]: done

File: pkg/operator/hostetcdendpointcontroller/members.go (hunk 1 above)
Comment (hexfusion) [reply pkg/operator/hostetcdendpointcontroller/members.go]: put an error for now but we need to think about how we handle this.

File: pkg/operator/hostetcdendpointcontroller/members.go (hunk 2 above)
Comment (hexfusion) [reply pkg/operator/hostetcdendpointcontroller/members.go]: done

**Comment by hexfusion** [comment]
@alaypatel07 updated PTAL

**Comment by hexfusion** [comment]
/hold 

adding IP address for bootstrap pod

**Comment by hexfusion** [comment]
> Container test exited with code 2, reason OOMKilled

odd
/test unit

File: pkg/operator/hostetcdendpointcontroller/members.go [hunk 5]
```diff
@@ -41,19 +43,31 @@ func (h *healthyEtcdMemberGetter) GetHealthyEtcdMembers() ([]string, error) {
...
+			return "", err
+		}
+		host, port, _ := net.SplitHostPort(u.Host)
+		//TODO peer port should be a global constant
+		if IsIP(host) && port == "2380" {
+			return "etcd-bootstrap", nil
 		}
 	}
-	return ""
+	return "", fmt.Errorf("peerURL %s is not properly formatted", peerURLs[0])
```
Comment (alaypatel07) [inline pkg/operator/hostetcdendpointcontroller/members.go]: s/peerURL/peerURLs?

File: pkg/operator/hostetcdendpointcontroller/members.go [hunk 6]
```diff
@@ -163,3 +177,12 @@ func (h *healthyEtcdMemberGetter) Endpoints() ([]string, error) {
 
 	return endpoints, nil
 }
+
+//TODO add to util
+func IsIP(addr string) bool {
+	ip := net.ParseIP(addr)
```
Comment (alaypatel07) [inline pkg/operator/hostetcdendpointcontroller/members.go]: Can we please use the variable in the if block below?

File: pkg/operator/hostetcdendpointcontroller/members.go (hunk 5 above)
Comment (hexfusion) [reply pkg/operator/hostetcdendpointcontroller/members.go]: that was intended as it was in single context `peerURLs[0]`

File: pkg/operator/hostetcdendpointcontroller/members_test.go [hunk 7]
```diff
@@ -24,19 +24,19 @@ func Test_getHostname(t *testing.T) {
...
-			args: args{peerURLs: []string{"https://etcd-0.foouser.tests.com"}},
+			args: args{peerURLs: []string{"https://etcd-0.foouser.tests.com:2380"}},
 			want: "etcd-0",
 		},
 		{
 			name: "valid test case for etcd bootstrap node",
-			args: args{peerURLs: []string{"https://etcd-bootstrap.foouser.tests.com"}},
+			args: args{peerURLs: []string{"https://10.0.139.142:2380"}},
 			want: "etcd-bootstrap",
 		},
```
Comment (alaypatel07) [inline pkg/operator/hostetcdendpointcontroller/members_test.go]: Can we add couple of more test cases for errors? malformed url and empty peerURLs?

File: pkg/operator/hostetcdendpointcontroller/members.go (hunk 5 above)
Comment (alaypatel07) [reply pkg/operator/hostetcdendpointcontroller/members.go]: This case will be hit only if the function does not return in the loop right? so it should be something like

`return "", fmt.Errorf("peerURLs %s is not properly formatted or empty", peerURLs)`

File: pkg/operator/hostetcdendpointcontroller/members_test.go (hunk 7 above)
Comment (hexfusion) [reply pkg/operator/hostetcdendpointcontroller/members_test.go]: sure done

File: pkg/operator/hostetcdendpointcontroller/members.go (hunk 6 above)
Comment (hexfusion) [reply pkg/operator/hostetcdendpointcontroller/members.go]: done

File: pkg/operator/hostetcdendpointcontroller/members.go (hunk 5 above)
Comment (hexfusion) [reply pkg/operator/hostetcdendpointcontroller/members.go]: changed the function a little as we don't need to loop at all as we only handle [0]. this makes the function more explicit and error handling is more clear.

File: pkg/operator/hostetcdendpointcontroller/members.go [hunk 8]
```diff
@@ -41,19 +43,33 @@ func (h *healthyEtcdMemberGetter) GetHealthyEtcdMembers() ([]string, error) {
 	}
 	hostnames := make([]string, 0)
 	for _, m := range member {
-		hostname := getEtcdName(m.PeerURLS)
+		hostname, err := getEtcdName(m.PeerURLS[0])
```
Comment (alaypatel07) [inline pkg/operator/hostetcdendpointcontroller/members.go]: Is it reasonable to assume that `len(m.PeerURLS) > 0`? if not this will be a nasty panic to catch

Comment (hexfusion) [reply pkg/operator/hostetcdendpointcontroller/members.go]: I think so we check that here https://github.com/openshift/cluster-etcd-operator/blob/f14a3303f3ff4807ffc3f8bd888b15786008e592/pkg/operator/hostetcdendpointcontroller/members.go#L109

Comment (alaypatel07) [reply pkg/operator/hostetcdendpointcontroller/members.go]: Perfect, thanks!
